src/converters/app-converter.ts

```diff
@@ -0,0 +1,388 @@
+import { mapToAwsResource } from '../aws-resource-mappings';
+import { attributePropertyName, mapToCfnResource } from '../cfn-resource-mappings';
+import { CloudFormationResource, getDependsOn } from '../cfn';
+import { OutputMap, OutputRepr } from '../output-map';
+import { parseSub } from '../sub';
+
+/**
+ * AppConverter will convert all CDK resources into Pulumi resources.
+ */
+export class AppConverter {
+    // Map of stack artifactId to StackConverter
+    public readonly stacks = new Map<string, StackConverter>();
+
+    public readonly manifestReader: AssemblyManifestReader;
+
+    constructor(readonly host: IStackComponent) {
+        this.manifestReader = AssemblyManifestReader.fromPath(host.assemblyDir);
+    }
+
+    convert() {
+        for (const stackManifest of this.manifestReader.stackManifests) {
+            const stackConverter = new StackConverter(this.host, stackManifest);
+            this.stacks.set(stackManifest.id, stackConverter);
+            this.convertStackManifest(stackManifest);
+        }
+    }
+
+    private convertStackManifest(artifact: StackManifest): void {
+        const dependencies = new Set<ArtifactConverter>();
+        for (const file of artifact.fileAssets) {
+            const converter = new FileAssetManifestConverter(this.host, file);
+            converter.convert();
+            dependencies.add(converter);
+        }
+
+        // TODO add docker asset converter
+        // for (const image of artifact.dockerAssets) {
+        // }
+
+        const stackConverter = this.stacks.get(artifact.id);
+        if (!stackConverter) {
+            throw new Error(`missing CDK Stack for artifact ${artifact.id}`);
+        }
+        stackConverter.convert(dependencies);
+    }
+}
+
+/**
+ * StackConverter converts all of the resources in a CDK stack to Pulumi resources
+ */
+export class StackConverter extends ArtifactConverter {
+    readonly parameters = new Map<string, any>();
+    readonly resources = new Map<string, Mapping<pulumi.Resource>>();
+    readonly constructs = new Map<ConstructInfo, pulumi.Resource>();
+    stackResource!: CdkConstruct;
+
+    constructor(host: IStackComponent, readonly stack: StackManifest) {
+        super(host);
+    }
+
+    public convert(dependencies: Set<ArtifactConverter>) {
+        const dependencyGraphNodes = GraphBuilder.build(this.stack);
+
+        // process parameters first because resources will reference them
+        for (const [logicalId, value] of Object.entries(this.stack.parameters ?? {})) {
+            this.mapParameter(logicalId, value.Type, value.Default);
+        }
+
+        for (const n of dependencyGraphNodes) {
+            if (n.construct.id === this.stack.id) {
+                this.stackResource = new CdkConstruct(`${this.host.name}/${n.construct.path}`, n.construct.id, {
+                    parent: this.host,
+                    // TODO: we could do better here. Currently the stack depends on all assets, but really
```

Comment:
Do we have an issue for this?

Does that mean that stacks depending on each other are serialized?

Reply:
So currently I don't think we are handing asset dependencies at all. The stack depends on the asset, but none of the resources depend on the stack so none of the resources depend on the asset. I think it ends in a race condition where asset publishing happens to be faster than anything else. 

Initially we can add a dependency  on all resources to depend on the stack. Ideally we would have individual resources depend on individual assets, but CDK doesn't track dependencies at this level only at the stack level.